core/src/main/java/org/apache/calcite/avatica/util/SecurityUtils.java

```diff
@@ -0,0 +1,266 @@
+    MethodHandles.Lookup lookup = MethodHandles.lookup();
+    try {
+      // Subject.getSubject(AccessControlContext) is deprecated for removal and replaced by
+      // Subject.current().
+      // Lookup first the new API, since for Java versions where both exists, the
+      // new API delegates to the old API (for example Java 18, 19 and 20).
+      // Otherwise (Java 17), lookup the old API.
+      return lookup.findStatic(Subject.class, "current",
+        MethodType.methodType(Subject.class));
+    } catch (Throwable x) {
+      try {
+        // This is a bit awkward, but the code is more symmetrical this way
+        return lookup.findStatic(SecurityUtils.class, "getSubjectFallback",
+          MethodType.methodType(Subject.class));
+      } catch (Throwable t) {
+        return null;
+      }
+    }
+  }
+
+  private static MethodHandle lookupGetSubject() {
+    MethodHandles.Lookup lookup = MethodHandles.lookup();
+    try {
+      Class<?> contextklass =
+          ClassLoader.getSystemClassLoader()
+              .loadClass("java.security.AccessControlContext");
+      return lookup.findStatic(Subject.class, "getSubject",
+        MethodType.methodType(Subject.class, contextklass));
+    } catch (Throwable t) {
+      return null;
+    }
+  }
+
+  private static MethodHandle lookupGetContext() {
+    try {
+      // Use reflection to work with Java versions that have and don't have AccessController.
+      Class<?> controllerKlass =
+          ClassLoader.getSystemClassLoader().loadClass("java.security.AccessController");
+      Class<?> contextklass =
+          ClassLoader.getSystemClassLoader()
+              .loadClass("java.security.AccessControlContext");
+
+      MethodHandles.Lookup lookup = MethodHandles.lookup();
+      return lookup.findStatic(controllerKlass, "getContext",
+        MethodType.methodType(contextklass));
+    } catch (Throwable x) {
+      return null;
+    }
+  }
+
+  /**
+   * Get the current security manager, if available.
+   * @return the current security manager, if available, null otherwise
+   */
+  public static Object getSecurityManager() {
+    try {
+      // Use reflection to work with Java versions that have and don't have SecurityManager.
+      return System.class.getMethod("getSecurityManager").invoke(null);
+    } catch (Throwable ignored) {
+      return null;
+    }
+  }
+
+  /**
+   * <p>
+   * Checks the given permission, if the {@link #getSecurityManager() security manager} is set.
+   * </p>
+   * @param permission the permission to check
+   * @throws SecurityException if the permission check fails
+   */
+  public static void checkPermission(Permission permission) throws SecurityException {
+    Object securityManager = SecurityUtils.getSecurityManager();
+    if (securityManager == null) {
+      return;
+    }
+    try {
+      securityManager.getClass().getMethod("checkPermission").invoke(securityManager,
+          permission);
+    } catch (SecurityException x) {
```

Comment:
I think this is wrong... it's being invoked via reflection so you'll need to catch `InvocationTargetException` and examine the contents.

Reply:
Yes, this is a straight bug.

Reply:
Should be fixed by unwrapInvocationTargetException.